### PR TITLE
Revision of hooks, fix of embedding normalization

### PR DESCRIPTION
**Pull request by rgemulla**
This integrates and revises the changes from #124 and #119 into a single PR. I.e.:

1. The hook system is greatly simplified. There are now pre/post hooks for batches, epochs, and run methods.
2. The hooks are used to fix a bug in embedder normalization, which is currently not applied correctly (see #119).

### CODE REVIEW

**Review by samuelbroscheit — COMMENTED** [review]
Looks good to me. Just two minor comments.

File: kge/job/entity_ranking.py [hunk 1]
```diff
@@ -387,41 +414,35 @@ def merge_hist(target_hists, source_hists):
...
-        trace_entry = self.trace(**trace_entry, echo=True, echo_prefix="  ", log=True)
+        # run hooks (may modify trace)
+        for f in self.post_epoch_hooks:
+            f(self)
+
+        # output the trace, then clear it
+        trace_entry = self.trace(
+            **self.current_trace["epoch"], echo=True, echo_prefix="  ", log=True
+        )
+        self.current_trace["epoch"] = None
```
[the remark this answers is not in the document]
Comment (rgemulla) [reply kge/job/entity_ranking.py:439]: Yes, but it only occurs twice and I found it clearer this way

File: kge/job/entity_ranking.py [hunk 2]
```diff
@@ -107,9 +106,40 @@ def _run(self) -> dict:
...
+        # run pre-epoch hooks (may modify trace)
+        for f in self.pre_epoch_hooks:
+            f(self)
+
         # let's go
         epoch_time = -time.time()
         for batch_number, batch_coords in enumerate(self.loader):
+            # create initial batch trace (yet incomplete)
+            self.current_trace["batch"] = dict(
+                type="entity_ranking",
```
[the remark this answers is not in the document]
Comment (rgemulla) [reply kge/job/entity_ranking.py:129]: Yes, it's different. job.type is traced as "job". We should keep this as it is like this in all our traces.

**Comment by rgemulla** [comment]
@samuelbroscheit ready to merge?

**Comment by samuelbroscheit** [comment]
Alright.

**Pull request merged**